### PR TITLE
Add SMPTE ST 2084 transfer function

### PR DESCRIPTION
**Pull request by mrd489**
SMPTE ST 2084 transfer function added in similar format to Rec 709 transfer function.

### CODE REVIEW

**Comment by sekrit-twc** [comment]
Some remarks:

Small things first:
1) Where are the tests?
2) Increment the API minor version

Bigger issues:
The main reason I haven't implemented ST2084 yet is the fact that it's an absolute function (0 and 1.0 have defined physical brightness) as opposed to the relative gamma-type functions. Simply converting between them as st2084(inv_709(x)) would make the assumption that 709 content was also coded for 0-10,000 cd/m^2, which is probably untrue. In order to support this, we need to introduce the notion of a nominal brightness level for relative transfer functions (gamma), and perhaps allow the caller to specify it at the API level.

File: src/zimg/colorspace/colorspace.h [hunk 1]
```diff
@@ -31,7 +31,8 @@ enum class MatrixCoefficients {
 enum class TransferCharacteristics {
 	TRANSFER_UNSPECIFIED,
 	TRANSFER_LINEAR,
-	TRANSFER_709
+	TRANSFER_709,
+	TRANSFER_2084
```
Comment (sekrit-twc) [inline src/zimg/colorspace/colorspace.h]: TRANSFER_ST2084

File: src/zimg/api/zimg.cpp [hunk 2]
```diff
@@ -221,13 +221,14 @@ zimg::colorspace::TransferCharacteristics translate_transfer(zimg_transfer_chara
...
-	static const zimg::static_map<zimg_transfer_characteristics_e, TransferCharacteristics, 6> map{
+	static const zimg::static_map<zimg_transfer_characteristics_e, TransferCharacteristics, 7> map{
 		{ ZIMG_TRANSFER_709,         TransferCharacteristics::TRANSFER_709 },
 		{ ZIMG_TRANSFER_UNSPECIFIED, TransferCharacteristics::TRANSFER_UNSPECIFIED },
 		{ ZIMG_TRANSFER_601,         TransferCharacteristics::TRANSFER_709 },
+		{ ZIMG_TRANSFER_LINEAR,      TransferCharacteristics::TRANSFER_LINEAR },
 		{ ZIMG_TRANSFER_2020_10,     TransferCharacteristics::TRANSFER_709 },
 		{ ZIMG_TRANSFER_2020_12,     TransferCharacteristics::TRANSFER_709 },
-		{ ZIMG_TRANSFER_LINEAR,      TransferCharacteristics::TRANSFER_LINEAR },
+		{ ZIMG_TRANSFER_2084,        TransferCharacteristics::TRANSFER_2084 },
```
Comment (sekrit-twc) [inline src/zimg/api/zimg.cpp]: ZIMG_TRANSFER_ST2084

**Comment by mrd489** [comment]
I overlooked the test directory and wrote an example function using the ST 2084 transfer function, so one still needs to be written.

Before submitting the pull request, I did consider the absolute luminance issue of ST 2084 that you described. Yes, this would not allow you to switch content between the Rec. 709 and ST 2084 transfer functions. But, this is similar to the way that you would not use zimg for color gamut remapping.

A reason that including the ST 2084 transfer function may be useful is for image resizing. In this case, it is important to first convert the gamma encoded values to linear light, then resize the image, then reapply the transfer function. Without converting back to linear light, dark image values are often crushed.

**Comment by sekrit-twc** [comment]
> Yes, this would not allow you to switch content between the Rec. 709 and ST 2084 transfer functions. But, this is similar to the way that you would not use zimg for color gamut remapping.

I disagree with the premise here. It has always been possible in z.lib to extend the gamut, e.g. from Rec.709 and Rec.2020. In fact, even the reverse operation is supported on floating-point data types (you get super-saturation). Likewise, conversions from LDR to HDR should yield meaningful results in z.lib.

**Comment by mrd489** [comment]
> Likewise, conversions from LDR to HDR should yield meaningful results in z.lib.

They absolutely should. When making the color gamut remapping, I was referring to using zimg as a color management system, which is not its intended use. A color management system would be ideal to preserve artistic intent. Likewise, specifying absolute luminance mappings for the Rec. 709 transfer function and ST 2084 would also be an artistic choice, and may not be as simple as specifying a single nominal brightness level.

While the above problem does need to be addressed, what about the concern with image resizing? Image resizing should be performed with linear light values. Therefore, wouldn't the ST 2084 transfer and inverse functions be necessary?

**Comment by sekrit-twc** [comment]
For performance reasons, z.lib does not automatically convert to linear when resizing (it would involve many steps). I agree that the ability to invert ST.2084 to avoid non-linearity artifacts when resizing is important, but all features should be fully-baked before commit.

Regarding artistic intent, dynamic range reduction indeed requires some subjective decisions to be made, but range increase from LDR to HDR should be a well-defined mechanical process, just as with gamut expansion. This is the main concern I have with the patch in its current state.

**Comment by mrd489** [comment]
Since zimg doesn't convert to linear values for resizing, then I do agree that it isn't ready for ST 2084 in this patch. There is more work to be done.

**Comment by sekrit-twc** [comment]
See PR #53.

**Pull request closed without merging**